subsys/fs/nvs/nvs.c

```diff
@@ -0,0 +1,836 @@
+		/* operation over sector boundary */
+		return -1;
+	}
+	return 0;
+}
+
+void _nvs_addr_advance(struct nvs_fs *fs, off_t *addr, u16_t step)
+{
+	*addr += step;
+	if (*addr >= (fs->sector_count * fs->sector_size)) {
+		*addr -= (fs->sector_count * fs->sector_size);
+	}
+}
+
+/* Getting the sector header for address given in offset */
+int _nvs_sector_hdr_get(struct nvs_fs *fs, off_t offset,
+	struct _nvs_sector_hdr *sector_hdr)
+{
+	return flash_read(fs->flash_device,
+			fs->offset + (offset & ~(fs->sector_size - 1)),
+			sector_hdr,
+			_nvs_len_in_flash(fs, sizeof(*sector_hdr)));
+}
+
+/* Initializing sector by writing magic and status to sector header,
+ * set fs->write_address just after header offset should be pointing
+ * to the sector
+ */
+int _nvs_sector_init(struct nvs_fs *fs, off_t offset)
+{
+	struct _nvs_sector_hdr sector_hdr;
+	int rc;
+
+	/* just to be sure, align offset to sector_size */
+	offset &= ~(fs->sector_size - 1);
+	fs->sector_id += 1;
+	rc = _nvs_sector_hdr_get(fs, offset, &sector_hdr);
+	if (rc) {
+		return NVS_ERR_FLASH;
+	}
+	if (sector_hdr.fd_magic != 0xFFFFFFFF) {
+		return NVS_ERR_NOSPACE;
+	}
+	sector_hdr.fd_magic = fs->magic;
+	sector_hdr.fd_id = fs->sector_id;
+	sector_hdr._pad = 0;
+	rc = nvs_flash_write(fs,
+		offset,
+		&sector_hdr,
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr)));
+	if (rc) {
+		return NVS_ERR_FLASH;
+	}
+	fs->write_location = offset +
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
+	return NVS_OK;
+}
+
+/* check if a sector is used by looking at all elements of the sector.
+ * return 0 if sector is empty, 1 if it is not empty
+ */
+int _nvs_sector_is_used(struct nvs_fs *fs, off_t offset)
+{
+	int rc;
+	off_t addr;
+	u8_t buf;
+
+	rc = 0;
+	offset &= ~(fs->sector_size - 1);
+	for (addr = 0; addr < fs->sector_size; addr += sizeof(buf)) {
+		rc = flash_read(fs->flash_device, fs->offset + addr,
+			&buf, sizeof(buf));
+		if (rc) {
+			return NVS_ERR_FLASH;
+		}
+		if (buf != 0xFF) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+/* erase the flash starting at offset, erase length = len */
+int _nvs_flash_erase(struct nvs_fs *fs, off_t offset, size_t len)
+{
+	int rc;
+
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return NVS_ERR_FLASH;
+	}
+
+	rc = flash_erase(fs->flash_device, fs->offset + offset, len);
+	if (rc) {
+		/* flash write error */
+		return NVS_ERR_FLASH;
+	}
+
+	SYS_LOG_DBG("Erasing flash at %x, len %x", offset, len);
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	return NVS_OK;
+}
+
+void _nvs_entry_sector_advance(struct nvs_fs *fs)
+{
+	fs->entry_sector++;
+	if (fs->entry_sector == fs->sector_count) {
+		fs->entry_sector = 0;
+	}
+}
+
+/* garbage collection: addr is set to the start of the sector to be gc'ed,
+ * the entry sector has been updated to point to the sector just after the
+ * sector being gc'ed
+ */
+int _nvs_gc(struct nvs_fs *fs, off_t addr)
```

Comment:
As I understand it, this algorithm is quadratic in the number of entries, even a mere 50 entries will result in at least 2500 reads (memcpy). I have some concerns over its performance, given also that it is ran automatically, and especially for the use case of nvs as a backend for the settings subsystem. @Laczen , what is your opinion on this?

Reply:
@lemrey Could you please elaborate this a bit more? Are this 50 different entries ? How many sectors,... What are your concerns over the performance?

Reply:
@Laczen Thanks for your reply. To elaborate, I had in mind a scenario with three sectors with 25 entries in each of the first two. I haven't thought about the number of unique entries, but I am unsure whether that would impact the number of reads. My concern is that GC can become very demanding: 2500 headers are ~10 KBytes, each GC iteration. My understanding is that it might be necessary to GC more than one sector to free up enough space for a given write operation.

Reply:
@lemrey I agree with your concern, I am not sure about the numbers: suppose that we take your scenario with the 25 entries in the first two, what happens in gc:
item 1: read header in gc sector, go through 25 entries in the second, read and write data
item 2: read header in gc sector, go through 25 entries +1 in the second, read and write data
item 3: ...
So in total: 25*(sum(1..25)+25)=25*(13+25)=950 including the 25 data reads. So 925 header reads and 25 data reads. Or 3.6kB + data reads.

GC will always be a demanding operation. In case of a solution with a "sticky" bit it will be less demanding, but a write operation would require a lot more reads to set the "sticky" bit.

Anyhow don't forget that GC has to remain a "rare" operation: each sector can only be GC'ed about 10k times, so each sector should only by GC'ed less than once a day to get an acceptable lifetime.

Reply:
Hi @lemrey, thanks for the comment, I value your comments a lot.

You are right, it does more reads, and your calculation is correct.

My analysis of the sticky bit also has a flaw: NVS also does the read when trying to write a new item to disk. It does this to avoid writing the same value twice. This could be avoided when a higher level interface (e.g. the settings module) filters out unnecessary writes (the settings module does this). Maybe a configuration setting (CONFIG_NVS_FLASH_PROTECT) to select the check against rewriting data should be added. What's your opinion?

Reply:
Hello, I believe it would be best to let another interface (or the application) check if the data has changed before attempting to write it to flash; from what I can see the settings subsystem already does that.

Reply:
@lemrey I have modified the NVS module, the check is now enabled by a configuration parameter (`CONFIG_NVS_PROTECT_FLASH`), by default it is disabled.